File: src/ui/js/ExportCharacter.js
/**
 *  Eternity Keeper, a Pillars of Eternity save game editor.
 *  Copyright (C) 2015 the authors.
 *
 *  Eternity Keeper is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  Eternity Keeper is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var ExportCharacter = function () {
	var self = this;

	var defaultState = {
		enabled: false
		, exporting: false
	};

	var populateSelector = characters => {
		var selector = self.html.exportCharacterDialog.find('.multi-selector');
		selector.empty();
		selector.append(characters.map(character =>
			$('<li>')
				.data('guid', character.GUID)
				.text(character.name)
				.append(' <i class="fa fa-spinner fa-pulse"></i>')
				.click(self.exportCharacter.bind(self, character.GUID))));
	};

	self.state = $.extend({}, defaultState);
	self.html = {};

	self.openDialog = () => {
		self.html.exportCharacterDialog.find('.alert').hide();
		populateSelector(Eternity.SavedGame.state.saveData.characters);
		self.html.exportCharacterDialog.modal('show');
	};

	self.exportCharacter = (guid, e) => {
		if (self.state.exporting) {
			return;
		}

		var spinner = $(e.currentTarget).find('i');

		var success = () => {
			self.state.exporting = false;
			spinner.hide();
			self.html.exportCharacterDialog.find('.alert-success').show();
		};

		var failure = (errno, response) => {
			self.state.exporting = false;
			spinner.hide();

			// NO_SAVENAME means the user cancelled the file dialog. The other
			// codes keep the dialog's own words; a reason given in words -- a
			// save that could only be read in part -- is shown as it stands.
			if (response !== 'NO_SAVENAME') {
				self.html.exportCharacterDialog.find('.alert-danger')
					.text(/^[A-Z_]+$/.test(response) ? 'Error exporting character.' : response)
					.show();
			}
		};

		self.html.exportCharacterDialog.find('.alert').hide();
		self.state.exporting = true;
		spinner.show();

		window.exportCharacter({
			request: JSON.stringify({
				GUID: guid
				, absolutePath: Eternity.SavedGame.state.info.absolutePath
				// An Apply or a Save moves the character's current state out
				// of the directory the list opened.
				, savedYet: Eternity.Modifications.state.savedYet
			})
			, onSuccess: success
			, onFailure: failure
		});
	};

	self.render = newState => {
		self.state = $.extend({}, defaultState, newState);
		var menuItem = self.html['menu-export-character'];
		menuItem.off();

		if (self.state.enabled) {
			menuItem.parent().removeClass('disabled');
			menuItem.click(self.openDialog);
		} else {
			menuItem.parent().addClass('disabled');
		}
	};
};

$.extend(ExportCharacter.prototype, Renderer.prototype);

File: src/main/java/uk/me/mantas/eternity/handlers/ExportCharacter.java
/**
 *  Eternity Keeper, a Pillars of Eternity save game editor.
 *  Copyright (C) 2015 the authors.
 *
 *  Eternity Keeper is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  Eternity Keeper is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


package uk.me.mantas.eternity.handlers;

import org.cef.browser.CefBrowser;
import org.cef.callback.CefQueryCallback;
import org.cef.handler.CefMessageRouterHandlerAdapter;
import org.json.JSONException;
import org.json.JSONObject;
import uk.me.mantas.eternity.Logger;
import uk.me.mantas.eternity.environment.Environment;
import uk.me.mantas.eternity.save.CharacterExporter;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;

import static org.cef.handler.CefDialogHandler.FileDialogMode;

public class ExportCharacter extends CefMessageRouterHandlerAdapter {
	private static final Logger logger = Logger.getLogger(ExportCharacter.class);

	@Override
	public boolean onQuery (
		final CefBrowser browser
		, final long id
		, final String request
		, final boolean persistent
		, final CefQueryCallback callback) {

		ChrDialog.choose(browser, FileDialogMode.FILE_DIALOG_SAVE, "Save Character"
			, callback, "NO_SAVENAME", filename -> export(request, filename, callback));

		return true;
	}

	@Override
	public void onQueryCanceled (final CefBrowser browser, final long id) {
		logger.error("Query #%d cancelled.%n", id);
	}

	private static void export (
		final String request, final String filename, final CefQueryCallback callback) {

		try {
			final JSONObject json = new JSONObject(request);
			final File save = Environment.getInstance().state().workingSave().forReading(
				new File(json.getString("absolutePath")), json.optBoolean("savedYet", false));

			final boolean exported = new CharacterExporter(save.getAbsolutePath()
				, json.getString("GUID"), ChrDialog.withChrExtension(filename)).export();

			if (exported) {
				callback.success("true");
			} else {
				callback.failure(-1, "EXPORT_ERR");
			}
		} catch (final JSONException e) {
			logger.error("Error parsing JSON request: %s%n", request);
			callback.failure(-1, "BAD_REQUEST");
		} catch (final FileNotFoundException e) {
			logger.error("Unable to find file : %s%n", e.getMessage());
			callback.failure(-1, "FILE_NOT_FOUND");
		} catch (final IOException e) {
			logger.error("Filesystem error: %s%n", e.getMessage());
			callback.failure(-1, "FILESYSTEM_ERR");
		}
	}
}
